### PR TITLE
Explain torch compile error messages, improve PTQ vs QAT doc

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4497
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4387
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4487
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4498

### CODE REVIEW

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
thanks ! But I don't think it closes https://github.com/zama-ai/concrete-ml-internal/issues/4414 as this issue is more about the error message itself, not the doc

**Comment by RomanBredehoft** [comment]
also might be worth to consider https://github.com/zama-ai/concrete-ml-internal/issues/4498 in this PR as well, if possible 🙏

**Review by yuxizama — COMMENTED** [review]
I suggest the following format for users to easily find solutions when they encounter some errors. Wdyt?

### Common Compilation erros 

#### Bit width limits
**Error message**: `This is the error message`
**Cause:** This error can occur when rounding_threshold_bits is not used and accumulated intermediate values in the computation exceed 16-bits. TLU input maximum bit-width is exceeded
**Possible solutions**: 
- Reduce quantization n_bits. However, this may reduce accuracy. When quantization n_bits must be below 6, it is best to use [Quantization Aware Training](https://github.com/zama-ai/concrete-ml/blob/docs/discuss_error_messages/docs/deep-learning/fhe_friendly_models.md).
- Use rounding_threshold_bits. This feature is described [here](https://github.com/zama-ai/concrete-ml/blob/docs/discuss_error_messages/docs/explanations/advanced_features.md#rounded-activations-and-quantizers). It is recommended to use the [fhe.Exactness.APPROXIMATE](https://github.com/zama-ai/concrete-ml/blob/docs/discuss_error_messages/docs/references/api/concrete.ml.torch.compile.md#function-compile_torch_model) setting, and set the rounding bits to 1 or 2 bits higher than the quantization n_bits
- Use [pruning](https://github.com/zama-ai/concrete-ml/blob/docs/discuss_error_messages/docs/explanations/pruning.md)

#### Crypto-parameters
**Error message:** `This is the error message`
**Cause:** This is the cause
**Possible sulutions:**
- Solution 1 
- Solution 2


#### Quantization
**Error message:** `This is the error message`
**Cause:** This is the cause
**Possible sulutions:**
- Solution 1 
- Solution 2

**Review by RomanBredehoft — DISMISSED** [review]

File: docs/deep-learning/torch_support.md [hunk 1]
```diff
@@ -100,7 +129,7 @@ In this example, the input values `x_test` and the predicted values `y_pred` are
 
 ## Simulated FHE Inference in the clear
 
-The user can also perform the inference on clear data. Two approaches exist:
+One can perform the inference on clear data in order to evaluate the impact of quantization and of FHE computation on the accuracy of their model. See [this section](../deep-learning/fhe_assistant.md#simulation) for more details. Two approaches exist:
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:132]:
```suggestion
You can perform the inference on clear data in order to evaluate the impact of quantization and of FHE computation on the accuracy of their model. See [this section](../deep-learning/fhe_assistant.md#simulation) for more details. Two approaches exist:
```

File: docs/deep-learning/fhe_assistant.md [hunk 2]
```diff
@@ -53,12 +53,62 @@ concrete_clf.fit(X, y)
 concrete_clf.compile(X, debug_config)
 ```
 
-## Compilation error debugging
+## Common compilation errors
 
-Compilation errors that signal that the ML model is not FHE compatible are usually of two types:
+The most common compilation errors stem from the following causes:
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:58]:
```suggestion

```

File: docs/deep-learning/fhe_assistant.md [hunk 3]
```diff
@@ -53,12 +53,62 @@ concrete_clf.fit(X, y)
...
 
-1. TLU input maximum bit-width is exceeded
-1. No crypto-parameters can be found for the ML model: `RuntimeError: NoParametersFound` is raised by the compiler
+#### 1. TLU input maximum bit-width is exceeded
+
+**Error message**: `this [N]-bit value is used as an input to a table lookup`
+
+**Cause**: This error can occur when `rounding_threshold_bits` is not used and accumulated intermediate values in the computation exceed 16-bits. The most common approaches to fix this issue are:
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:64]:
```suggestion
**Cause**: This error can occur when `rounding_threshold_bits` is not used and accumulated intermediate values in the computation exceed 16 bits. 
```

File: docs/deep-learning/fhe_assistant.md [hunk 4]
```diff
@@ -53,12 +53,62 @@ concrete_clf.fit(X, y)
...
+
+**Possible solutions**:
+
+- Reduce quantization `n_bits`. However, this may reduce accuracy. When quantization `n_bits` must be below 6, it is best to use [Quantization Aware Training](../deep-learning/fhe_friendly_models.md).
+- Use `rounding_threshold_bits`. This feature is described [here](../explanations/advanced_features.md#rounded-activations-and-quantizers). It is recommended to use the [`fhe.Exactness.APPROXIMATE`](../references/api/concrete.ml.torch.compile.md#function-compile_torch_model) setting, and set the rounding bits to 1 or 2 bits higher than the quantization `n_bits`
+- Use [pruning](../explanations/pruning.md)
+
+#### 2. No crypto-parameters can be found
+
+**Error message**: `RuntimeError: NoParametersFound` is raised by the compiler
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:74]:
```suggestion
**Error message**: `RuntimeError: NoParametersFound`
```

File: docs/deep-learning/fhe_assistant.md [hunk 5]
```diff
@@ -53,12 +53,62 @@ concrete_clf.fit(X, y)
...
+
+**Cause**: This error occurs when using `rounding_threshold_bits` in the `compile_torch_model` function.
+
+**Possible solutions**: The solutions in this case are similar to the ones for the previous error.
+
+#### 3. Quantization import failed
+
+**Error message**: `Error occurred during quantization aware training (QAT) import [...] Could not determine a unique scale for the quantization!`.
+
+**Cause**: This error is a due to missing quantization operators in the model that is imported as a quantized aware training model. See [this guide](../deep-learning/fhe_friendly_models.md) on how to use Brevitas layers. This error message is generated when not all layers take inputs that are quantized through `QuantIdentity` layers.
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:84]:
```suggestion
**Cause**: This error occurs when the model imported as a quantized-aware training model lacks quantization operators. See [this guide](../deep-learning/fhe_friendly_models.md) on how to use Brevitas layers. This error message indicates that some layers do not take inputs quantized through `QuantIdentity` layers.
```

File: docs/deep-learning/fhe_assistant.md [hunk 6]
```diff
@@ -53,12 +53,62 @@ concrete_clf.fit(X, y)
...
+
+<!--pytest-codeblocks:skip-->
+
+```python
+z = torch.cat([self.quant_concat(x), self.quant_concat(y)])
+```
+
+## Debugging compilation errors
+
+Compilation errors due to FHE incompatible models, such as maximum bit-width exceeded or `NoParametersFound` can be debugged by examining the bit-widths associated with various intermediate values of the FHE computation.
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:111]:
```suggestion
For compilation errors due to FHE-incompatible models, such as maximum bit-width exceeded or `NoParametersFound`, you can debug them by examining the bit-widths associated with various intermediate values of the FHE computation.
```

File: docs/deep-learning/fhe_assistant.md [hunk 7]
```diff
@@ -116,6 +166,8 @@ Function you are trying to compile cannot be compiled:
 
 The error `this 17-bit value is used as an input to a table lookup` indicates that the 16-bit limit on the input of the Table Lookup (TLU) operation has been exceeded. To pinpoint the model layer that causes the error, Concrete ML provides the [bitwidth_and_range_report](../references/api/concrete.ml.quantization.quantized_module.md#method-bitwidth_and_range_report) helper function. First, the model must be compiled so that it can be [simulated](fhe_assistant.md#simulation).
 
+On the other hand, `NoParametersFound` is encountered when using `rounding_threshold_bits`. When using this setting, the 16-bit accumulator limit is relaxed. However, reducing bit-width, or reducing the `rounding_threshold_bits`, or using  using the [`fhe.Exactness.APPROXIMATE`](../references/api/concrete.ml.torch.compile.md#function-compile_torch_model) rounding method can help.
```
Comment (yuxizama) [inline docs/deep-learning/fhe_assistant.md:169]:
```suggestion
On the other hand, `NoParametersFound` occurs when using `rounding_threshold_bits`. With this setting, the 16-bit accumulator limit is relaxed. However, the following solution could help: 
- Reduce bit width
- Reduce the `rounding_threshold_bits`
- Use the [`fhe.Exactness.APPROXIMATE`](../references/api/concrete.ml.torch.compile.md#function-compile_torch_model) rounding method.
```

File: docs/deep-learning/torch_support.md [hunk 8]
```diff
@@ -2,14 +2,21 @@
...
 
-As [Quantization Aware Training (QAT)](../explanations/quantization.md) is the most appropriate method of training neural networks that are compatible with [FHE constraints](../getting-started/concepts.md#model-accuracy-considerations-under-fhe-constraints), Concrete ML works with [Brevitas](../explanations/inner-workings/external_libraries.md#brevitas), a library providing QAT support for PyTorch.
+There are two approaches to build [FHE-compatible deep networks](../getting-started/concepts.md#model-accuracy-considerations-under-fhe-constraints):
 
-The following example uses a simple QAT PyTorch model that implements a fully connected neural network with two hidden layers. Due to its small size, making this model respect FHE constraints is relatively easy.
+- [Quantization Aware Training (QAT)](../explanations/quantization.md) requires using custom layers, but can quantize weights and activations to low bit-widths. Concrete ML works with [Brevitas](../explanations/inner-workings/external_libraries.md#brevitas), a library providing QAT support for PyTorch. To use this mode, compile models using `compile_brevitas_qat_model`
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:7]:
```suggestion
- [**Quantization Aware Training (QAT)**](../explanations/quantization.md) : QAT requires using custom layers, but can quantize weights and activations to low bit widths. Concrete ML works with [Brevitas](../explanations/inner-workings/external_libraries.md#brevitas), a library providing QAT support for PyTorch. To use this mode, compile models using `compile_brevitas_qat_model`
```

File: docs/deep-learning/torch_support.md [hunk 9]
```diff
@@ -2,14 +2,21 @@
...
+- Post-training Quantization: in this mode a vanilla PyTorch model can be compiled. However, when quantizing weights & activations to fewer than 7 bits the accuracy can decrease strongly. On the other hand, depending on the model size, quantizing with 6-8 bits can be incompatible with FHE constraints. To use this mode, compile models with `compile_torch_model`.
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:8]:
```suggestion
- **Post-training Quantization**: This mode allows a vanilla PyTorch model to be compiled. However, when quantizing weights & activations to fewer than 7 bits, the accuracy can decrease strongly. On the other hand, depending on the model size, quantizing with 6-8 bits can be incompatible with FHE constraints. To use this mode, compile models with `compile_torch_model`.
```

File: docs/deep-learning/torch_support.md [hunk 10]
```diff
@@ -2,14 +2,21 @@
...
+
+Both approaches should be used with the `rounding_threshold_bits` parameter set accordingly. The best values for this parameter need to be determined through experimentation. A good initial value to try is `6`. See [here](../explanations/advanced_features.md#rounded-activations-and-quantizers) for more details.
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:10]:
```suggestion
Both approaches require the `rounding_threshold_bits` parameter to be set accordingly. The best values for this parameter need to be determined through experimentation. A good initial value to try is `6`. See [here](../explanations/advanced_features.md#rounded-activations-and-quantizers) for more details.
```

File: docs/deep-learning/torch_support.md [hunk 11]
```diff
@@ -51,38 +58,60 @@ torch_model = QATSimpleNet(30)
...
 
-Using the templates above, the probability of obtaining the target accumulator bit-width, for a single layer, was determined experimentally by training 10 models for each of the following data-sets.
+## Configuring quantization parameters
 
-| <p>probability of obtaining<br>the accumulator bit-width</p> | 8   | 10   | 12  | 14  | 16   |
-| ------------------------------------------------------------ | --- | ---- | --- | --- | ---- |
-| mnist,fashion                                                | 72% | 100% | 72% | 85% | 100% |
-| cifar10                                                      | 88% | 88%  | 75% | 75% | 88%  |
-| cifar100                                                     | 73% | 88%  | 61% | 66% | 100% |
+With QAT, the PyTorch/Brevitas models, created following the example above, require the user to configure quantization parameters such as `bit_width` (activation bit-width) and `weight_bit_width`. When using this mode, set `n_bits=None` in the `compile_brevitas_qat_model`.
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:110]:
```suggestion
With QAT (the PyTorch/Brevitas models created following the example above), you need to configure quantization parameters such as `bit_width` (activation bit-width) and `weight_bit_width`. When using this mode, set `n_bits=None` in the `compile_brevitas_qat_model`.
```

File: docs/deep-learning/torch_support.md [hunk 12]
```diff
@@ -51,38 +58,60 @@ torch_model = QATSimpleNet(30)
...
 
-Note that the accuracy on larger data-sets, when the accumulator size is low, is also reduced strongly.
+With PTQ, the user needs to set the `n_bits` value in the `compile_torch_model` function. A trade-off between accuracy and FHE compatibility and latency must be determined manually.
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:112]:
```suggestion
With PTQ, you need to set the `n_bits` value in the `compile_torch_model` function and must manually determine the trade-off between accuracy, FHE compatibility, and latency.
```

File: docs/deep-learning/torch_support.md [hunk 13]
```diff
@@ -51,38 +58,60 @@ torch_model = QATSimpleNet(30)
...
 
-| <p>accuracy for target<br>accumulator bit-width</p> | 8   | 10  | 12  | 14  | 16  |
-| --------------------------------------------------- | --- | --- | --- | --- | --- |
-| cifar10                                             | 20% | 37% | 89% | 90% | 90% |
-| cifar100                                            | 6%  | 30% | 67% | 69% | 69% |
+The quantization parameters, along with the number of neurons on each layer, will determine the accumulator bit-width of the network. Larger accumulator bit-widths result in higher accuracy but slower FHE inference time.
```
Comment (yuxizama) [inline docs/deep-learning/torch_support.md:114]:
```suggestion
The quantization parameters, along with the number of neurons on each layer, will determine the accumulator bit width of the network. Larger accumulator bit widths result in higher accuracy but slower FHE inference time.
```

**Review by RomanBredehoft — DISMISSED** [review]

**Review by RomanBredehoft — APPROVED** [review]

**Review by RomanBredehoft — APPROVED** [review]

**Pull request merged**